klippy/webhooks.py

```diff
@@ -0,0 +1,245 @@
+        for req in requests:
+            logging.debug(
+                "ServerConnection: Request received from Moonraker %s" % (req))
+            try:
+                decoded_req = json_loads_byteified(req)
+                self._process_request(decoded_req)
+            except Exception:
+                logging.exception(
+                    "ServerConnection: Error processing Server Request %s"
+                    % (req))
+
+    def _process_request(self, req):
+        web_request = WebRequest(req)
+        try:
+            func = self.webhooks.get_callback(
+                web_request.get_path())
+            func(web_request)
+        except WebRequestError as e:
+            web_request.set_error(e)
+        except Exception as e:
+            web_request.set_error(WebRequestError(e.message))
+        result = web_request.finish()
+        logging.debug(
+            "ServerConnection: Sending response - %s" % (str(result)))
+        self.send({'method': "response", 'params': result})
+
+    def send(self, data):
+        if not self.is_server_connected:
+            return
+        with self.mutex:
+            retries = 10
+            data = json.dumps(data) + "\x00"
+            while data:
+                try:
+                    sent = self.socket.send(data)
+                except socket.error as e:
+                    if e.errno == 9 or e.errno == 32 or not retries:
+                        sent = 0
+                    else:
+                        retries -= 1
+                        waketime = self.reactor.monotonic() + .001
+                        self.reactor.pause(waketime)
+                        continue
+                retries = 10
+                if sent > 0:
+                    data = data[sent:]
+                else:
+                    logging.info(
+                        "ServerConnection: Error sending server data,"
+                        " closing socket")
+                    self.close_socket()
+                    break
+
+class WebHooks:
+    def __init__(self, printer):
+        self._endpoints = {}
+        self._hooks = []
+        self.sconn = ServerConnection(self, printer)
+
+    def register_endpoint(self, path, callback, methods=['GET'], params={}):
+        if path in self._endpoints:
+            raise WebRequestError("Path already registered to an endpoint")
+
+        methods = [m.upper() for m in methods]
+        for method in methods:
+            if method not in AVAILABLE_METHODS:
+                raise WebRequestError(
+                    "Requested Method [%s] for endpoint '%s' is not valid"
+                    % (method, path))
+
+        self._endpoints[path] = callback
+        hook = (path, methods, params)
+        self._hooks.append(hook)
+        # Send Hook to server if connected
+        if self.sconn.is_connected():
+            self.sconn.send(
+                {'method': "add_klippy_hook", 'params': {'hook': hook}})
```

Comment:
FWIW, could we instead make a "list endpoints" endpoint to query all the available methods?